### ISSUE TITLE
performance: UDP recvmmsg instead of recvfrom

### ISSUE THREAD

**Issue by splitice** [vectordotdev/vector#12452]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Particularly when running in VMs with mitigations (e.g those CPUs used at many popular cloud vendors...) syscalls can be rather expensive.

### Attempted Solutions

N/A

### Proposal

Use recvmmsg instead of recvfrom for udp message receiving for improved performance.

With some of our applications pushing 100k messages at peak per log message performance is very important.

### References

Relates to my performance drive in #12451

### Version

0.21.0

**Comment by jszwedko** [comment]
Hi @splitice !

Thanks for opening this.

I might be misunderstanding, but wouldn't each of `recvfrom` or `recvmsg` result in a single syscall? I think we use `recvfrom` because we want the address of the sender to include as event metadata.

**Comment by jszwedko** [comment]
Aaah, sorry, I misread `recvmmsg` as `recvmsg`. I see now.

**Comment by splitice** [comment]
@jszwedko yes the syscall allows for the receiving of multiple messages at the time (and includes the receive from address)

I'm not sure if there is a neat way to access it from rust or a crate for plaform independant access i.e windows IOCP and similar.

This was the closest issue I found https://github.com/tokio-rs/mio/issues/185

When scrolling through a github search for "tokio recvmmsg" I did find numerous people who built their own support, so it would appear that would be necessary as tokio doesnt provide vectorized IO (e.g recvmmg) support.

e.g https://github.com/sebastiencs/rustorrent/blob/6829a99fb94a4badc948533f4fc75a9f0e4a32c2/src/utp/udp_socket.rs
